File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        // unordered_set<vector<int>> st;
        vector<vector<int>> res;
        for (int i = 0; i < n; ++i) {
            if (i != 0 && nums[i] == nums[i-1]) continue;
            int left = i+1;
            int right = n-1;
            int target = -nums[i];
            while (left < right) {
                int curr_sum = nums[left] + nums[right];
                if (curr_sum == target) {
                    res.push_back({nums[i], nums[left], nums[right]});
                    while (left < right && nums[left] == nums[left+1]) {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right-1]) {
                        right--;
                        
                    }
                    left++;
                    right--;
                }
                else if (curr_sum < target) {
                    left++;
                }
                else right--;
            }
        }
        // for (auto a : st) res.push_back(a);
        return res;
    }
};

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        int len = nums.length;
        List<List<Integer>> res = new ArrayList<>();
        // Run loop only for 
        for (int i = 0; i < len - 2; ++i) {

            int start = i + 1;
            int end = len - 1; 
            while (start < end) {
                if (nums[i] + nums[start] + nums[end] == 0) {
                    res.add(Arrays.asList(nums[i], nums[start], nums[end]));
                    // Handle inside of {start, end} as well because they could also be duplicates. 
                    while (start < end && nums[start] == nums[start + 1]) start++;
                    while (start < end && nums[end] == nums[end - 1]) end--;
                    start++;
                    end--;
                }
                else if (nums[i] + nums[start] + nums[end] > 0) {
                    end--;
                }
                else {
                    start++;
                }
            }

            // To be able to remove the duplicates, we should post process the i and skip for the same next element.
            // If we pre process this, then we will lose the first occurence of that i itself. 
            while (i < len-1 && nums[i] == nums[i+1]) {
                i++;
            }
        }
        return res;
    }
}